cmake/compiler.cmake

```diff
@@ -0,0 +1,28 @@
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_C_STANDARD 11)
+
+##! Global Configuration
+set(COMMON_CXX_FLAGS "-Wall -Wextra -Wfatal-errors -Wno-unknown-pragmas")
+
+##! Configuration for debug builds
+
+macro(__internal_debug_cxx_flags)
+    set(CMAKE_CXX_FLAGS_DEBUG "-O0 -g3 ${COMMON_CXX_FLAGS}")
+endmacro()
+
+##! Configuration for release builds
+macro(__internal_release_cxx_flags)
+    set(CMAKE_CXX_FLAGS_RELEASE "-O3 -march=native ${COMMON_CXX_FLAGS}")
+endmacro()
+
+##! Compiler Configuration
+if (CMAKE_BUILD_TYPE MATCHES Debug)
```

Comment:
I recently learn that's we don't have to use Conditional on CMake Build Type but generator expression

Reply:
Can you elaborate on that / link some resources ? I don't know that feature

Reply:
https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html

```CMake
## for example -> $<$<CONFIG:DEBUG>:FOO_EXTRA_THINGS>
set(CMAKE_CXX_FLAGS "$<$<AND:$<CONFIG:Release>,$<CXX_COMPILER_ID:CLANG>>:-O3>"
```

Reply:
from CMake documentation -> Generator expressions are evaluated during build system generation to produce information specific to each build configuration.

Reply:
Honestly, I think regular conditions are easier on the eyes ^^